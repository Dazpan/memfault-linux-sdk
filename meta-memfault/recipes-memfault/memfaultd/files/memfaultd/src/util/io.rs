//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::io::BufReader;

/// A trait for getting the length of a stream.
/// Note std::io::Seek also has a stream_len() method, but that method is fallible.
pub trait StreamLen {
    /// Gets the length of the stream in bytes.
    fn stream_len(&self) -> u64;
}

impl<R: StreamLen> StreamLen for BufReader<R> {
    fn stream_len(&self) -> u64 {
        self.get_ref().stream_len()
    }
}
