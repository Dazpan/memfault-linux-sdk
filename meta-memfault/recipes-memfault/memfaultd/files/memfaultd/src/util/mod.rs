//
// Copyright (c) Memfault, Inc.
// See License.txt for details
pub mod fs;
pub mod io;
pub mod path;
#[cfg(feature = "logging")]
pub mod rate_limiter;
pub mod serialization;
pub mod string;
pub mod system;
pub mod task;
pub mod zip;
