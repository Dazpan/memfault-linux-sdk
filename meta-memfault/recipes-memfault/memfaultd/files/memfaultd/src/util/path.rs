//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::ffi::OsStr;
use std::path::Path;

/// Splits the filename at the first dot.
/// This is similar to the nighly-only std::path::Path::file_prefix.
pub fn file_prefix(path: &Path) -> Option<&OsStr> {
    let file_name = path.file_name()?;
    file_name.to_str()?.split('.').next().map(OsStr::new)
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case("/test", "test")]
    #[case("/test.log", "test")]
    #[case("/test.log.zlib", "test")]
    fn test_file_prefix(#[case] path: &str, #[case] expected: &str) {
        assert_eq!(file_prefix(Path::new(path)), Some(OsStr::new(expected)));
    }
}
