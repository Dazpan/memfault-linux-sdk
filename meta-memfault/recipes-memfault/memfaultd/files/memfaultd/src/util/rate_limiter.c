//! @file
//!
//! Copyright (c) Memfault, Inc.
//! See License.txt for details
//!
//! @brief
//! Rate limiter library functions

#include "memfault/util/rate_limiter.h"

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

#include "memfaultd.h"

struct MemfaultdRateLimiter {
  int count;
  int duration;
  time_t *history;
  char *file;
};

bool memfaultd_rate_limiter_check_event(sMemfaultdRateLimiter *handle) {
  if (!handle) {
    //! Rate limiting disabled
    return true;
  }

  struct timeval now;
  gettimeofday(&now, NULL);

  if (handle->history[handle->count - 1] + handle->duration > now.tv_sec) {
    fprintf(stderr, "rate_limiter:: Rejecting event, rate limit reached\n");
    return false;
  }

  //! Shuffle right, adding new time into [0]
  for (int i = handle->count - 2; i >= 0; --i) {
    handle->history[i + 1] = handle->history[i];
  }
  handle->history[0] = now.tv_sec;

  if (handle->file) {
    FILE *fd = fopen(handle->file, "w");
    if (!fd) {
      //! Failed to write rate_limit file, but return true to proceed with rest of processing
      fprintf(stderr, "rate_limiter:: Failed to open rate_limit file\n");
      return true;
    }

    for (int i = 0; i < handle->count; ++i) {
      fprintf(fd, "%ld ", handle->history[i]);
    }

    fclose(fd);
  }

  return true;
}

void memfaultd_rate_limiter_destroy(sMemfaultdRateLimiter *handle) {
  if (handle) {
    free(handle->file);
    free(handle->history);
    free(handle);
  }
}

sMemfaultdRateLimiter *memfaultd_rate_limiter_init(sMemfaultd *memfaultd, const int count,
                                                   const int duration, const char *filename) {
  if (count == 0 || duration == 0 || !memfaultd) {
    return NULL;
  }

  sMemfaultdRateLimiter *handle;
  if (!(handle = calloc(sizeof(sMemfaultdRateLimiter), 1))) {
    fprintf(stderr, "rate_limiter:: Failed to allocate handle\n");
    goto cleanup;
  }

  handle->count = count;
  handle->duration = duration;

  if (!(handle->history = calloc(sizeof(time_t), handle->count))) {
    fprintf(stderr, "rate_limiter:: Failed to allocate history array\n");
    goto cleanup;
  }

  if (filename) {
    if (!(handle->file = memfaultd_generate_rw_filename(memfaultd, filename))) {
      fprintf(stderr, "rate_limiter:: Failed to generate history filename\n");
      goto cleanup;
    }

    FILE *fd = fopen(handle->file, "r");
    if (!fd) {
      if (errno != ENOENT) {
        //! File exists, but we can't open it
        fprintf(stderr, "rate_limiter:: Failed to open history file\n");
        goto cleanup;
      }
    } else {
      for (int i = 0; i < handle->count; ++i) {
        if (fscanf(fd, "%ld ", &handle->history[i]) != 1) {
          handle->history[i] = 0;
          break;
        }
      }

      fclose(fd);
    }
  }

  return handle;

cleanup:

  free(handle->file);
  free(handle->history);
  free(handle);
  return NULL;
}

#ifdef MEMFAULT_UNITTEST

time_t *memfaultd_rate_limiter_get_history(sMemfaultdRateLimiter *handle) {
  return handle->history;
}

#endif