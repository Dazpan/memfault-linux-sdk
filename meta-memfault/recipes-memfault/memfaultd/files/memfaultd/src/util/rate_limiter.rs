//
// Copyright (c) Memfault, Inc.
// See License.txt for details
//! Support for rate limiting some execution paths.
use eyre::Result;
use ratelimit::Ratelimiter;

/// A rate limiter which keeps track of how many calls were rate limited.
/// You can also provide a `info: I` parameter with each call. The latest one
/// will be passed to the runner when not rate limited anymore.
pub struct RateLimiter<I> {
    rate_limiter: Ratelimiter,
    limited_calls: Option<RateLimitedCalls<I>>,
}

pub struct RateLimitedCalls<I> {
    pub count: usize,
    pub latest_call: I,
}

impl<I> RateLimiter<I> {
    /// Create a new rate limiter with given capacity, quantum and rate (see ratelimit::Ratelimiter).
    pub fn new(capacity: u64, quantum: u64, rate: u64) -> Self {
        Self {
            rate_limiter: Ratelimiter::new(capacity, quantum, rate),
            limited_calls: None,
        }
    }

    /// Run the provided work function if the rate limitings limits have not been reached.
    pub fn run_within_limits<W>(&mut self, info: I, work: W) -> Result<()>
    where
        W: FnOnce(Option<RateLimitedCalls<I>>) -> Result<()>,
    {
        if self.rate_limiter.try_wait().is_ok() {
            work(self.limited_calls.take())
        } else {
            self.limited_calls = Some(match self.limited_calls.take() {
                None => RateLimitedCalls {
                    count: 1,
                    latest_call: info,
                },
                Some(l) => RateLimitedCalls {
                    count: l.count + 1,
                    latest_call: info,
                },
            });
            Ok(())
        }
    }
}
