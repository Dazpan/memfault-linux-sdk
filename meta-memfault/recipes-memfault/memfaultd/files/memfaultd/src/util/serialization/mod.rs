//
// Copyright (c) Memfault, Inc.
// See License.txt for details
pub mod kib_to_usize;
pub mod milliseconds_to_duration;
pub mod seconds_to_duration;
