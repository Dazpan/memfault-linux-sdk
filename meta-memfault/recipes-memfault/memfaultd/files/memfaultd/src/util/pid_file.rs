//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::fs::OpenOptions;
use std::io::{ErrorKind, Write};
use std::path::Path;
use std::process::id;

use eyre::{Report, Result, WrapErr};
use nix::unistd::Pid;

const PID_FILE: &str = "/var/run/memfaultd.pid";

pub fn write_memfaultd_pid_file() -> Result<()> {
    let file = OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(Path::new(PID_FILE));

    match file {
        Ok(mut file) => writeln!(file, "{}", id()).wrap_err("Failed to write PID file"),
        Err(e) => {
            let msg = match e.kind() {
                ErrorKind::AlreadyExists => "Daemon already running, aborting.",
                _ => "Failed to open PID file, aborting.",
            };
            Err(Report::new(e).wrap_err(msg))
        }
    }
}

pub fn get_pid_from_file() -> Result<Pid> {
    match std::fs::read_to_string(PID_FILE) {
        Ok(pid_string) => {
            let pid = pid_string
                .trim()
                .parse()
                .wrap_err("Failed to parse PID file contents")?;
            Ok(Pid::from_raw(pid))
        }
        Err(e) => {
            eprintln!("Unable to read memfaultd PID file.");
            Err(e.into())
        }
    }
}
