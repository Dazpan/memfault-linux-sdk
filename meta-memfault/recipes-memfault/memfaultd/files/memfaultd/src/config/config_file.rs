//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use serde::{Deserialize, Serialize};
use std::time::Duration;
use std::{net::SocketAddr, path::PathBuf};

use crate::util::serialization::*;
use crate::util::*;

#[derive(Serialize, Deserialize, Debug)]
pub struct MemfaultdConfig {
    #[serde(rename = "queue_size_kib", with = "kib_to_usize")]
    pub queue_size: usize,
    pub data_dir: PathBuf,
    #[serde(rename = "refresh_interval_seconds", with = "seconds_to_duration")]
    pub refresh_interval: Duration,
    pub enable_data_collection: bool,
    pub enable_dev_mode: bool,
    pub software_version: String,
    pub software_type: String,
    pub project_key: String,
    pub base_url: String,
    pub swupdate_plugin: SwUpdateConfig,
    pub reboot_plugin: RebootPlugin,
    pub collectd_plugin: CollectdPlugin,
    pub coredump_plugin: CoredumpPlugin,
    #[serde(rename = "fluent-bit")]
    pub fluent_bit: FluentBitConfig,
    pub logs: LogsConfig,
    pub mar: MarConfig,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct SwUpdateConfig {
    pub input_file: PathBuf,
    pub output_file: PathBuf,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct RebootPlugin {
    pub last_reboot_reason_file: PathBuf,
    pub uboot_fw_env_file: PathBuf,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct CollectdPlugin {
    pub header_include_output_file: PathBuf,
    pub footer_include_output_file: PathBuf,
    pub non_memfaultd_chain: String,
    #[serde(rename = "write_http_buffer_size_kib", with = "kib_to_usize")]
    pub write_http_buffer_size: usize,
    #[serde(rename = "interval_seconds", with = "seconds_to_duration")]
    pub interval: Duration,
}

#[derive(Serialize, Deserialize, Debug)]
pub enum CoredumpCompression {
    #[serde(rename = "gzip")]
    Gzip,
    #[serde(rename = "none")]
    None,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct CoredumpPlugin {
    pub compression: CoredumpCompression,
    #[serde(rename = "coredump_max_size_kib", with = "kib_to_usize")]
    pub coredump_max_size: usize,
    pub rate_limit_count: u32,
    #[serde(rename = "rate_limit_duration_seconds", with = "seconds_to_duration")]
    pub rate_limit_duration: Duration,
    #[serde(rename = "storage_min_headroom_kib", with = "kib_to_usize")]
    pub storage_min_headroom: usize,
    #[serde(rename = "storage_max_usage_kib", with = "kib_to_usize")]
    pub storage_max_usage: usize,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct FluentBitConfig {
    pub extra_fluentd_attributes: Vec<String>,
    pub bind_address: SocketAddr,
    pub max_buffered_lines: usize,
    pub max_connections: usize,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct LogsConfig {
    #[serde(rename = "rotate_size_kib", with = "kib_to_usize")]
    pub rotate_size: usize,

    #[serde(rename = "rotate_after_seconds", with = "seconds_to_duration")]
    pub rotate_after: Duration,

    pub tmp_folder: PathBuf,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct MarConfig {
    #[serde(rename = "storage_max_usage_kib", with = "kib_to_usize")]
    pub storage_max_usage: usize,
}

use serde_json::Value;
use std::fs;
use std::path::Path;

impl MemfaultdConfig {
    const DEFAULT_CONFIG_PATH: &str = "/etc/memfaultd.conf";

    pub fn load(config_path: Option<&Path>) -> eyre::Result<MemfaultdConfig> {
        // Initialize with the builtin config file.
        let mut config: Value = Self::parse(include_str!("../../../libmemfaultc/builtin.conf"))?;

        // Select config file to read
        let user_config_path = config_path.unwrap_or_else(|| Path::new(Self::DEFAULT_CONFIG_PATH));

        // Read and parse the user config file.
        let user_config = Self::parse(std::fs::read_to_string(user_config_path)?.as_str())?;

        // Merge the two JSON objects together
        Self::merge_into(&mut config, user_config);

        // Load the runtime config but only if the file exists. (Missing runtime config is not an error.)
        let runtime_config_path = Self::generate_runtime_config_path(&config)?;
        if runtime_config_path.exists() {
            let runtime = Self::parse(fs::read_to_string(runtime_config_path)?.as_str())?;
            Self::merge_into(&mut config, runtime);
        }

        // Transform the JSON object into a typed structure.
        Ok(serde_json::from_value(config)?)
    }

    // Parse a Memfaultd JSON configuration file (with optional C-style comments) and return a serde_json::Value object.
    fn parse(config_string: &str) -> eyre::Result<Value> {
        let json_text = string::remove_comments(config_string);
        let json: Value = serde_json::from_str(json_text.as_str())?;
        if !json.is_object() {
            return Err(eyre::eyre!("Configuration should be a JSON object."));
        }
        Ok(json)
    }

    /// Merge two JSON objects together. The values from the second one will override values in the first one.
    fn merge_into(dest: &mut Value, src: Value) {
        assert!(dest.is_object() && src.is_object());
        if let Value::Object(dest_map) = src {
            for (key, value) in dest_map {
                if let Some(obj) = dest.get_mut(&key) {
                    if obj.is_object() {
                        MemfaultdConfig::merge_into(obj, value);
                        continue;
                    }
                }
                dest[&key] = value;
            }
        }
    }

    // Generate the path to the runtime config file from a serde_json::Value object. This should include the "data_dir" field.
    fn generate_runtime_config_path(config: &Value) -> eyre::Result<PathBuf> {
        let mut data_dir = PathBuf::from(
            config["data_dir"]
                .as_str()
                .ok_or(eyre::eyre!("Config['data_dir'] must be a string."))?,
        );
        data_dir.push("runtime.conf");
        Ok(data_dir)
    }
}

#[cfg(test)]
mod test {
    use rstest::rstest;

    use super::*;

    #[test]
    fn test_merge() {
        let mut c =
            serde_json::from_str(r##"{ "node": { "value": true, "valueB": false } }"##).unwrap();
        let j = serde_json::from_str(r##"{ "node2": "xxx" }"##).unwrap();

        MemfaultdConfig::merge_into(&mut c, j);

        assert_eq!(
            serde_json::to_string(&c).unwrap(),
            r##"{"node":{"value":true,"valueB":false},"node2":"xxx"}"##
        );
    }

    #[test]
    fn test_merge_overwrite() {
        let mut c =
            serde_json::from_str(r##"{ "node": { "value": true, "valueB": false } }"##).unwrap();
        let j = serde_json::from_str(r##"{ "node": { "value": false }}"##).unwrap();

        MemfaultdConfig::merge_into(&mut c, j);

        assert_eq!(
            serde_json::to_string(&c).unwrap(),
            r##"{"node":{"value":false,"valueB":false}}"##
        );
    }

    #[test]
    fn test_merge_overwrite_nested() {
        let mut c = serde_json::from_str(
            r##"{ "node": { "value": true, "valueB": false, "valueC": { "a": 1, "b": 2 } } }"##,
        )
        .unwrap();
        let j = serde_json::from_str(r##"{ "node": { "valueC": { "b": 42 } }}"##).unwrap();

        MemfaultdConfig::merge_into(&mut c, j);

        assert_eq!(
            serde_json::to_string(&c).unwrap(),
            r##"{"node":{"value":true,"valueB":false,"valueC":{"a":1,"b":42}}}"##
        );
    }

    #[rstest]
    #[case("empty_object")]
    #[case("with_partial_logs")]
    #[case("without_coredump_compression")]
    fn can_parse_test_files(#[case] name: &str) {
        let input_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("src/config/test-config")
            .join(name)
            .with_extension("json");
        // Verifies that the file is parsable
        let content = MemfaultdConfig::load(Some(&input_path)).unwrap();
        // And that the configuration generated is what we expect.
        // Use `cargo insta review` to quickly approve changes.
        insta::assert_json_snapshot!(name, content)
    }
}
