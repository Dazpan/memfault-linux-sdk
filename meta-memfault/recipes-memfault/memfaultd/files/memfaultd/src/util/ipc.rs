//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use eyre::Result;
use nix::sys::signal::kill;
use nix::sys::signal::Signal;

use crate::util::pid_file::get_pid_from_file;

pub fn send_flush_queue_signal() -> Result<()> {
    let pid = get_pid_from_file()?;

    kill(pid, Signal::SIGUSR1)?;

    Ok(())
}
