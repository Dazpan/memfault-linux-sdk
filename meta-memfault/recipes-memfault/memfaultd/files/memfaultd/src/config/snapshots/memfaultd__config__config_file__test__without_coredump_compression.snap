---
source: memfaultd/src/config/config_file.rs
expression: content
---
{
  "queue_size_kib": 1024,
  "data_dir": "/media/memfault",
  "refresh_interval_seconds": 3600,
  "enable_data_collection": false,
  "enable_dev_mode": false,
  "software_version": "0.0.0-memfault-unknown",
  "software_type": "memfault-unknown",
  "project_key": "",
  "base_url": "https://device.memfault.com",
  "swupdate_plugin": {
    "input_file": "/etc/swupdate.cfg",
    "output_file": "/tmp/swupdate.cfg"
  },
  "reboot_plugin": {
    "last_reboot_reason_file": "/media/last_reboot_reason"
  },
  "collectd_plugin": {
    "header_include_output_file": "/tmp/collectd-header-include.conf",
    "footer_include_output_file": "/tmp/collectd-footer-include.conf",
    "non_memfaultd_chain": "",
    "write_http_buffer_size_kib": 64,
    "interval_seconds": 3600
  },
  "coredump_plugin": {
    "compression": "none",
    "coredump_max_size_kib": 96000,
    "rate_limit_count": 5,
    "rate_limit_duration_seconds": 3600,
    "storage_min_headroom_kib": 10240,
    "storage_max_usage_kib": 0
  },
  "fluent-bit": {
    "extra_fluentd_attributes": [],
    "bind_address": "127.0.0.1:5170",
    "max_buffered_lines": 1000,
    "max_connections": 4
  },
  "logs": {
    "rotate_size_kib": 10240,
    "rotate_after_seconds": 3600,
    "tmp_folder": "tmplogs",
    "compression_level": 1
  },
  "mar": {
    "storage_max_usage_kib": 102400,
    "mar_file_max_size_kib": 10240
  }
}
